blob/memstore/memstore.go

```diff
@@ -213,18 +213,17 @@ func (s *KV) Delete(_ context.Context, key string) error {
 }
 
 // List implements part of [blob.KV].
-func (s *KV) List(_ context.Context, start string, f func(string) error) error {
-	s.μ.Lock()
-	defer s.μ.Unlock()
-
-	for e := range s.m.InorderAfter(entry{key: start}) {
-		if err := f(e.key); errors.Is(err, blob.ErrStopListing) {
-			return nil
-		} else if err != nil {
-			return err
+func (s *KV) List(_ context.Context, start string) iter.Seq2[string, error] {
+	return func(yield func(string, error) bool) {
+		s.μ.Lock()
```

Comment:
Not a novel issue in this change, but: if this lock is held, does that mean looping over a list and issuing some Gets will deadlock?

I think this should be documented as part of the base store's required semantics: either all methods remain usable during a List iteration, or none are. IMO it shouldn't be left to the discretion of each store implementation, since that's effectively equivalent to forbidding other method calls during the ist.

Reply:
Yeah, several stores have this limitation. I have been kind of meaning to add some text to the effect that the implementation should document whether it's safe to mutate the collection during a list. And this change made me look at it again, but I still haven't fixed it 😅

Reply:
(Or possibly to say "you can't do that unless the implementation specifically says you may")

Reply:
Really, though, it's silly that `memstore` does not permit reading while listing, so I fixed that.